include/mppic/utils.hpp

```diff
@@ -125,6 +126,62 @@ auto shortest_angular_distance(
   return normalize_angles(to - from);
 }
 
+inline size_t findPathFurthestPoint(const models::CriticFunctionData & data)
+{
+  auto path_points = xt::view(data.path, xt::all(), xt::range(0, 2));
+
+  auto last_points_ext =
+    xt::view(data.trajectories, xt::all(), -1, xt::newaxis(), xt::range(0, 2));
+  auto distances = xt::norm_l2(last_points_ext - path_points, {2});
+  size_t max_id_by_trajectories = 0;
+  double min_distance_by_path = std::numeric_limits<double>::max();
+
+  for (size_t i = 0; i < distances.shape(0); i++) {
+    size_t min_id_by_path = 0;
+    for (size_t j = 0; j < distances.shape(1); j++) {
+      if (min_distance_by_path < distances(i, j)) {
+        min_distance_by_path = distances(i, j);
+        min_id_by_path = j;
+      }
+    }
+    max_id_by_trajectories = std::max(max_id_by_trajectories, min_id_by_path);
+  }
+  return max_id_by_trajectories;
+}
+
+
+inline void setPathFurthestPointIfNotSet(models::CriticFunctionData & data)
+{
+  if (!data.furthest_reached_path_point) {
+    data.furthest_reached_path_point = findPathFurthestPoint(data);
+  }
+}
+
+inline double distanceFromFurthestToGoal(const models::CriticFunctionData & data)
+{
+  if (!data.furthest_reached_path_point) {
+    throw std::runtime_error("Furthest point not computed yet");
+  }
+  auto path_points = xt::view(data.path, xt::all(), xt::range(0, 2));
+  auto furthest_reached_path_point = xt::view(
+    path_points, *data.furthest_reached_path_point,
+    xt::all());
+  auto goal_point = xt::view(path_points, -1, xt::all());
+
+  return xt::norm_l2(furthest_reached_path_point - goal_point, {0})();
+}
+
+inline double pathRatioReached(const models::CriticFunctionData & data)
```

Comment:
> But isn't this really a radius from goal thing? Thinking about where this is used:

Idea of this critics was to use them when we stuck, and trajectories just at the beginning of the path. Last goal can be near, but trajectories not cover path (imagine obstacle in front of robot, and goal right after, then the goal will be near, but our trajectories stuck in obstacle, and path goes right or left around obstacle, so we don't cover path, but we near the goal)

Reply:
> So even if we don't use distance from goal, even just using something like total integrated path length is a more natural metric

For this case it's not. If this metric will be in meters this would require to adjust this value to the total path length. Given just the ratio it's intuitive much more intuitive for what i think it should do. We activate critics only when path covered on say no more then 25%, and i don't need to think about meters, path length or something else. That was the idea, and it works as expected now.

With one amendment, that path_angle critic should be activated if robot heading heading back, so we don't need prefer forward critic again.

Reply:
OK, we can stick with this. The `obstacle in front of robot` situation the integrated path distance metric would handle. However, I think your metric also handles the stuck condition if we're stopped so all of the trajectories are relatively short, so that we kick it on to get started even if in close proximity to the goal. Is that right? If so, adding a comment about why this metric is a good metric would be valuable in the critic plugins so that future readers know what we were thinking. 

> With one amendment, that path_angle critic should be activated if robot heading back to the nearest path point, so we don't need prefer forward critic again.

I think we do from my testing, it still drove backwards for a full trajectory when I requested something behind it. See comment below where I outline that

Reply:
Because i didn't add condition yet

Reply:
added condition